File: unity/Assets/Sanguo/Contracts/GridLayout.cs
using System;
namespace Sanguo.Contracts
{
    // Dimensionless presentation coordinates, NOT path cost or attack rules.
    // Same fixtures as Java GridLayout; source-image origins are metadata, not an extra world offset.
    public sealed class GridLayout
    {
        public readonly bool ColumnStaggered;
        public readonly double Offset;
        public readonly int SourceOriginX, SourceOriginY;
        public GridLayout(bool columnStaggered,double offset,int originX,int originY)
        { ColumnStaggered=columnStaggered;Offset=offset;SourceOriginX=originX;SourceOriginY=originY; }
        public double X(int q,int r) { return ColumnStaggered?r:q+r*.5-Offset; }
        public double Z(int q,int r) { return ColumnStaggered?q+r*.5-Offset:r; }
        public int Row(double x,double z) { return (int)Math.Floor((ColumnStaggered?x:z)+.5); }
        public int Column(double x,double z) { return (int)Math.Floor((ColumnStaggered?z:x)-Row(x,z)*.5+Offset+.5); }
    }
}

File: game-api/src/main/java/game/sanguo/api/GridLayout.java
package game.sanguo.api;

/** Storage (q,r) to dimensionless presentation X/Z. Native source grids transpose axes.
 * Rendering can interpolate a continuous surface; this does not define movement costs. */
public final class GridLayout {
    public final boolean columnStaggered;
    public final double offset;
    public final int sourceOriginX, sourceOriginY;
    public GridLayout(boolean columnStaggered,double offset,int sourceOriginX,int sourceOriginY){
        this.columnStaggered=columnStaggered;this.offset=offset;
        this.sourceOriginX=sourceOriginX;this.sourceOriginY=sourceOriginY;
    }
    public double x(int q,int r){return columnStaggered?r:q+r*.5-offset;}
    public double z(int q,int r){return columnStaggered?q+r*.5-offset:r;}
    public int r(double x,double z){return (int)Math.floor((columnStaggered?x:z)+.5);}
    public int q(double x,double z){int r=r(x,z);return (int)Math.floor((columnStaggered?z:x)-r*.5+offset+.5);}
}
